### DISCUSSION TITLE
Explicit dependence on coordinates in class method

### DISCUSSION THREAD

**Discussion by Divyoj-Singh** [zwicker-group/py-pde#81] [Q&A]
I am trying to implement the following equations:
[pdes.pdf](https://github.com/zwicker-group/py-pde/files/6197742/pdes.pdf)
This is the code I wrote, mainly modifying the brusselator code:
The theta from the equation has been changed to variable z and is periodic in 2pi.

`import numba as nb
import numpy as np
import pde
import math as math
from pde import FieldCollection, PDEBase, PlotTracker, ScalarField, UnitGrid

class polarityPDE(PDEBase):
    """ Cell polarity """

    def __init__(self, alpha=1, length=0.5,gamma=0.1, diffusion_constant=0.5, bc=["periodic","periodic",'periodic']):
   
        self.alpha = alpha
        self.length = length
        self.Diffusion= diffusion_constant  # spatial mobility
        self.bc = bc  # boundary condition
        self.gamma=gamma,

    def get_initial_state(self, grid):
        """ preparing initial state """
        a = ScalarField(grid, 0.01, label="Field $a$")
        b = ScalarField(grid, 0.001,label="Field $b$")
        return FieldCollection([a, b])

    def function_f(self,value,gamma):
        if (value < 1/gamma):
            return value*gamma
        if (value > 1/gamma):
            return 1
    
    def evolution_rate(self, state, t=0):
        """ python implementation of the PDE """
        a, b = state
        rhs = state.copy()
        
        alp=self.alpha
        D=self.Diffusion
        l=self.length
        gam=self.gamma
        rhs[0] = alp*(5-a.project(['z'])/(2*math.pi)) - (1-self.function_f(b('x'+l*math.cos('z'),'y'+math.sin('z'),'z'+math.pi),gam))*a + D*a.laplace(self.bc)
        
        rhs[1] = alp*(5-b.project(['z'])/(2*math.pi)) - (1-self.function_f(a('x'+l*math.cos('z'),'y'+math.sin('z'),'z'+math.pi),gam))*b + D*b.laplace(self.bc)
        return rhs

# initialize state
        
T = 500.0;
L=2;

grid = pde.CartesianGrid([[0,L], [0, L],[0,2*math.pi]], [30, 30, 50],[False,False,True])  # generate grid

eq = polarityPDE(diffusion_constant=0.2)
state = eq.get_initial_state(grid)

# simulate the pde
tracker = PlotTracker(interval=1, plot_args={"vmin": 0, "vmax": 2})
sol = eq.solve(state, t_range=T, dt=1e-2, tracker=tracker)

### finally i want to plot the the value of a and b in x and y and z..`

but I can't understand how to access the value of a at  x, y and z and thus write the second term. Also, I am still searching how to get only the z component of laplace at the last term.

Can I please get help in this ?
Thanks!

**Comment by david-zwicker** [comment]
Unfortunately, I do not have the time to help people with implementing specific PDEs. However, I'm happy to help with specific questions considering the package.

**Reply by Divyoj-Singh** [reply]
I am really sorry for the trouble. I tried going back and forth between all the examples, tutorials, discussion etc. 
Still trying to get the hang of the library. 
The specific question I am trying to ask here is if we can implement pdes of type :
![image](https://user-images.githubusercontent.com/76565220/112483203-d1d9e780-8d9e-11eb-8e24-bdced60898d3.png)

a simplistic version of the above code is the following:
```
import numba as nb
import numpy as np
import pde
import math as math
from pde import FieldCollection, PDEBase, PlotTracker, ScalarField, UnitGrid

class polarityPDE(PDEBase):
    def __init__(self,  bc=["periodic","periodic",'periodic']):
           self.bc = bc  
    def get_initial_state(self, grid):
        a = ScalarField(grid, 0.01, label="Field $a$")
        return a
    
    def evolution_rate(self, state, t=0):
        a = state
        rhs = state.copy()
        l=1
        a_near= a('x'+l*math.cos('z'),'y'+l*math.sin('z'),'z'+math.pi)
        rhs=a_near
        return rhs

T = 500.0; L=2;
grid = pde.CartesianGrid([[0,L], [0, L],[0,2*math.pi]], [30, 30, 50],[True,True,True])  # generate grid
eq = polarityPDE()
state = eq.get_initial_state(grid)
sol = eq.solve(state, t_range=T)
```

This gives the expected error:
  a_near= a('x'+l*math.cos('z'),'y'+math.sin('z'),'z'+math.pi)
TypeError: must be real number, not str

But, I cant understand how to mitigate it.
Thanks again.

**Reply by david-zwicker** [reply]
This equation does not seem to be a partial differential equation, so it does not really fall into the scope of this package.

**Reply by Divyoj-Singh** [reply]
I am sorry for the confusion but I had removed other parts from the equations, as I need help in this part specifically.
The complete equation looks like this:
![image](https://user-images.githubusercontent.com/76565220/113417815-e3943e00-93e1-11eb-9ee1-838b6e0c739c.png)
This has partial derivative wrt to z as well. 
If there is a way to implement it in py-pde it would be a saver. I have tried various other solver packages but none of them provide this fine customisation. 
Implementing everything from scratch is really very difficult. 
Again, thanks for the help!

**Reply by david-zwicker** [reply]
I'm sorry, but this class of equations is beyond the scope of `py-pde`, mostly because it contains non-local terms.
